components/brave_today/browser/brave_news_p3a.cc

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2022 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "brave/components/brave_today/browser/brave_news_p3a.h"
+
+#include <algorithm>
+
+#include "base/cxx17_backports.h"
+#include "base/metrics/histogram_functions.h"
+#include "base/metrics/histogram_macros.h"
+#include "brave/components/brave_today/common/pref_names.h"
+#include "brave/components/weekly_storage/weekly_storage.h"
+#include "components/prefs/pref_registry_simple.h"
+#include "components/prefs/pref_service.h"
+
+namespace {
+
+uint16_t UpdateWeeklyStorageWithValueAndGetMax(PrefService* prefs,
+                                               const char* pref_name,
+                                               const uint64_t total) {
+  WeeklyStorage storage(prefs, pref_name);
+  storage.ReplaceTodaysValueIfGreater(total);
+  return storage.GetHighestValueInWeek();
+}
+
+uint64_t AddToWeeklyStorageAndGetSum(PrefService* prefs,
+                                     const char* pref_name,
+                                     int change) {
+  WeeklyStorage storage(prefs, pref_name);
+  if (change > 0)
+    storage.AddDelta(1);
+  return storage.GetWeeklySum();
+}
+
+template <std::size_t N>
+void RecordToHistogramBucket(const char* histogram_name,
+                             const int (&buckets)[N],
```

Comment:
Does this make `std::end()` faster?  I've had reviewers complain about templating before, because of code bloat, although it's probably fine in file-scope as here.